E2E: revise `ElementHelper.clickTab` method to be more resilient against lazy-loading elements.

worldomonation opened the pull request:
#### Proposed Changes

This PR revises the behavior of `ElementHelper.clickTab` with more modern (Playwright 1.14+) conventions, in order to add resilience to the `clickTab` method that is used by multiple POMs to interact with the navtabs.

Key changes:
- move up the short-circuiting behavior to be the first check in the method;
- move selectors to the selectors object;
- add an explicit wait for the `main` element to stabilize for mobile viewports only.

Details:
The current implementation of `ElementHelper.clickTab` has several shortcomings that allow flakiness to surface when interacting with pages containing navtabs that also have lazy loading elements. The issue is described in https://github.com/Automattic/wp-calypso/issues/64395.

By first checking for the currently active tab against the target tab and short circuiting if the two are equal, we are able to eliminate redundant clicks, saving a few milliseconds.

By using `Locator` instead of `ElementHandle`, we are able to encapsulate logic to locate the element when the lookup occurs, avoiding one potential area of issue should the layout shift.


#### Testing Instructions

Ensure the following:
  - [x] Gutenberg E2E (desktop)
  - [x] Gutenberg E2E (mobile)
  - [x] Calypso E2E (desktop)
  - [x] Calypso E2E (mobile)
  - [x] Pre-Release E2E

Related to https://github.com/Automattic/wp-calypso/issues/64395.
Closes https://github.com/Automattic/wp-calypso/issues/64389.

worldomonation committed:
Refactor

worldomonation commented on packages/calypso-e2e/src/element-helper.ts:
```diff
@@ -41,41 +47,54 @@ export async function waitForElementEnabled(
-		const navTabsButtonLocator = page.locator( selectors.mobileNavTabsToggle );
+		// Layout can shift in certain pages (eg. PlansPage) as elements are lazy loaded.
+		const mainElement = await page.waitForSelector( selectors.main );
+		await Promise.all( [
+			page.waitForLoadState( 'networkidle' ),
+			mainElement.waitForElementState( 'stable' ),
+		] );
```
This is the meat of the change; by waiting for the `main` element to stabilize, we can be reasonably certain that all loading has finished, even if `networkidle` has already fired.

The reason for using `main` is that not all pages that use the navtabs have the same element hierarchy.
For example, `PlansPage` has `div.plans` as the container, but in Users page the same element is only `div`.

dpasque replied:
So, I've spent a bit of time debugging this, and I'm not entirely sure that waiting for the 'stable' animation is going to _always_ stop this failure, although I think it will help.

I think Louis's comment here: p1654583626042029/1654264833.113049-slack-C1A1EKDGQ is explaining what is going on. There's a react lifecycle method that's firing at the end of everything that's closing the mobile pane.

I've watched this page load throttled several times, and here's what you usually see:
1. The top nav toolbar loads in with a loading indicator underneath it.
2. All the page plan information loads, but the SVGs/icons haven't loaded yet.
3. The last thing to happen is all the SVG/icons load in. ---> this is when you see the react lifecycle method fire too, and the mobile pane close.

The catch is that, unless I'm misunderstanding the docs, [all that "stable" is looking for under the hood](https://playwright.dev/docs/actionability#stable) is just for the bounding rectangle of an element to be the same dimensions over two consecutive re-paints/re-render frames.

The catch is that, from what I can tell playing around in the Chrome console, that's true for several frames _before_ that final load happens. Especially with how this is called in `Promise.all`, this happens several times right as the page loads, so that part of the promise will likely resolve really early.

Moving that stable check out of the `Promise.all` might slow things down enough that we'll be less likely to see the failure, but theoretically, it probably could still happen.

Since this is unique to the plans page, I wonder if we might be best off just putting in a one-off kind of hack for that page until the lifecycle issue is fixed?

Since the very last thing that seems to happen is that the Plan svg's load in, maybe we could do a wait for one of those responses in the POM for the plans page?

```typescript
await page.waitForResponse(response => response.url().includes("/images/wpcom-ecommerce"));
```

It's hacky, but I think it might be even more reliable, and it lets us isolate the hacky-ness to the single page where this is a problem.

I'm definitely not married to that idea though!

Thoughts?

worldomonation replied:
Yes, certainly good points!

> [all that "stable" is looking for under the hood](https://playwright.dev/docs/actionability#stable) is just for the bounding rectangle of an element to be the same dimensions over two consecutive re-paints/re-render frames...this happens several times right as the page loads, so that part of the promise will likely resolve really early

You are 100% correct. I have often resorted to using `stable` (especially in the early days of the framework) to try to deal with animations, but it's often an incomplete workaround for the reasons you state here.

If you recall, the lazy-loading upsell banner in the `SidebarComponent` also suffers from the same problem. From Playwright's perspective, the `SidebarComponent` is already stable even though in reality the upsell banner is just about to destroy this man's whole automation flow, which is what ended up happening for us.

> Moving that stable check out of the Promise.all might slow things down enough that we'll be less likely to see the failure, but theoretically, it probably could still happen.

Yeah, this would just end up being akin to a variable version of the `page.waitForTimeout` which is sketchy at best.

> Since this is unique to the plans page, I wonder if we might be best off just putting in a one-off kind of hack for that page until the lifecycle issue is fixed?

As much as you and (and everyone maintaining this down the road) will hate it, this may be the only option we have.

I think rest of the changes to `ElementHelper.clickTab` is worth keeping, so my proposal is this:
- remove the additional wait on `main` to be stable (restore the initial condition, which was only a `networkidle`);
- in `PlansPage.clickTab`, add a specific wait for an SVG as you suggest.

dpasque replied:
That proposal is great! And I agree - I thought the other improvements to the function were really nice! :)

worldomonation replied:
Pushed up f506b0d478179945ff25371efe1bf27f157b2d13 which implements the suggestions.

worldomonation commented on packages/calypso-e2e/src/element-helper.ts:
```diff
@@ -1,10 +1,16 @@
 	// clickNavTab
-	navTab: ( tab: string ) => `.section-nav-tab:has-text("${ tab }")`,
-	mobileNavTabsToggle: 'button.section-nav__mobile-header',
+	navTabItem: ( { name = '', selected }: { name?: string; selected?: boolean } = {} ) =>
+		`${ navTabParent } a[aria-current="${ selected }"]:has(span:has-text("${ name }"))`,
```
I've made this selector slightly more complex but this allows us to specify any four permutation of the input options and thus allowing us to avoid having to embed a selector within template literals (which we do in a few other places):

```typescript
`${ selectors.navTabItem( 'navtabname' ) }[aria-current="selected"]`
```

worldomonation committed:
Strip numerals from extracted tab name string to deal with tabs under PostsPage.

worldomonation commented on packages/calypso-e2e/src/element-helper.ts:
```diff
@@ -41,41 +47,57 @@ export async function waitForElementEnabled(
+	// Short circuit operation if the active tab and target tabs are the same.
+	// Strip numerals from the extracted tab name to account for the slightly
+	// different implementation in PostsPage.
+	if ( selectedTabName.replace( /[0-9]/g, '' ) === name ) {
```
This is seen in the later section of this file as well.

Within `PostsPage`, the navtabs display the number of posts that match the category denoted by the navtab:
![image](https://user-images.githubusercontent.com/6549265/172502141-d0ab7b7e-7e1a-4437-82f7-b25347476a5d.png)

Since there would be no situation (as of writing) that calls for an arabic numeral in as part of the navtab name, I felt that this was an acceptable call to make.

dpasque commented on packages/calypso-e2e/src/element-helper.ts:
```diff
@@ -1,10 +1,16 @@
 	// clickNavTab
-	navTab: ( tab: string ) => `.section-nav-tab:has-text("${ tab }")`,
-	mobileNavTabsToggle: 'button.section-nav__mobile-header',
+	navTabItem: ( { name = '', selected }: { name?: string; selected?: boolean } = {} ) =>
```
When not selected, do we want to make sure that we search for `[aria-current="false"]`? Because right now the default fallback to this would be undefined, which might not make the selector what we want.

worldomonation replied:
Right, that's a byproduct of me waffling between making it an optional or required selector. Will address in next push.

worldomonation replied:
Fixed in f506b0d478179945ff25371efe1bf27f157b2d13.

dpasque reviewed (commented):
A lot of the changes look good, but I have a discussion question about a possible alternative that may help catch this in even more cases. Let me know what you think!

worldomonation committed:
Revise approach based on PR feedback:
- remove wait on `main` element in ElementHelper;
- add wait for a SVG to load in the Legacy Plans grid in PlansPage;

worldomonation commented on packages/calypso-e2e/src/lib/pages/plans-page.ts:
```diff
@@ -153,25 +153,18 @@ export class PlansPage {
+			);
 		}
 
-		if ( targetTab === 'My Plan' ) {
-			// User is currently on the Plans tab and going to My Plans.
-			// Wait for the Plans grid to fully render.
-			const plansGridLocator = this.page.locator( selectors.legacyPlansGrid );
-			await plansGridLocator.waitFor();
-		}
-		if ( targetTab === 'Plans' ) {
-			// User is currently on the My Plans tab and going to Plans.
-			// Wait for the detais of the current plan to complete rendering
-			// asynchronously.
-			const placeholderLocator = this.page.locator( `.my-plan-card${ selectors.placeholder }` );
-			await placeholderLocator.waitFor( { state: 'hidden' } );
-		}
```
With the wait for SVG added in f506b0d478179945ff25371efe1bf27f157b2d13, it should be possible to remove the extra guards placed here.

dpasque reviewed (approved):
Looks awesome! 🚀  Here's hoping this does the trick! :D
